api.md

```diff
@@ -0,0 +1,392 @@
+to:
+
+```elixir
+  def create(conn, %{"image" => image}) do
+
+    # Check if file is an image
+    fileIsAnImage = String.contains?(image.content_type, "image")
+
+    if fileIsAnImage do
+
+      # Upload to S3
+      upload = image.path
+      |> ExAws.S3.Upload.stream_file()
+      |> ExAws.S3.upload("imgup-original", image.filename, acl: :public_read)
+      |> ExAws.request
+
+      # Check if upload was successful
+      case upload do
+        {:ok, _body} ->
+          render(conn, :success)
+
+        {:error, error} ->
+
+          {_error_atom, http_code, body} = error
+          render(conn |> put_status(http_code), body)
+      end
+
+    # If it's not an image, return 400
+    else
+      render(conn |> put_status(400), %{body: "File is not an image."})
+    end
+
+  end
+```
+
+Now we are checking the filetype of the uploaded file
+and providing feedback *back to the person*!
+
+
+# 4 Returning the URL to the person
+
+On success,
+it might be useful for the person using the API
+to get the public link where the image is stored.
+For this,
+we simply need to **parse the XML response from the `S3` upload**.
+For this, visit `lib/app_web/controllers/api_controller.ex`
+and change the `def create(conn, %{"image" => image})`
+to this:
+
+```elixir
+  def create(conn, %{"image" => image}) do
+
+    # Check if file is an image
+    fileIsAnImage = String.contains?(image.content_type, "image")
+
+    if fileIsAnImage do
+
+      # Upload to S3
+      upload = image.path
+      |> ExAws.S3.Upload.stream_file()
+      |> ExAws.S3.upload("imgup-original", image.filename, acl: :public_read)
+      |> ExAws.request
+
+      # Check if upload was successful
+      case upload do
+        {:ok, body} ->
+          url = body.body |> xpath(~x"//text()") |> List.to_string()
+          render(conn, :success, %{url: url})
+
+        {:error, error} ->
+
+          {_error_atom, http_code, body} = error
+          render(conn |> put_status(http_code), body)
+      end
+
+    # If it's not an image, return 400
+    else
+      render(conn |> put_status(400), %{body: "File is not an image."})
+    end
+
+  end
+```
+
+We are using
+[`sweet_xml`](https://github.com/kbrw/sweet_xml)
+we've imported earlier in our dependencies list.
+This will allow us to parse the output from the `S3` upload,
+which is in `XML` format.
+It has a `<Location>` tag,
+which is the URl we are interested in returning to the user.
+
+We then pass this URL to the `render/3` function.
+All we need to do is change it 
+to *return this URL to the person*.
+
+Open `lib/app_web/controllers/api_json.ex` 
+and change to.
+
+```elixir
+  def render("success.json", assigns) do
+    %{url: assigns.url}
+  end
+```
+
+Awesome!
+Now if you run `mix phx.server`
+and make a `multipart/form-data` request
+(we recommend using a tool like 
+[`Postman`](https://www.postman.com/) or [`Hoppscotch`](https://hoppscotch.io/) - which is open-source!),
```

Comment:
@LuchoTurtle you were clearly _using_ `Hoppscotch` to test this on your `localhost` when you were writing it. 
Why can't you include an actual set of instructions with screenshots for how to test it so that the reader doesn't have to spend _ages_ figuring it out for themself? 🤷‍♂️

Reply:
Reading: https://docs.hoppscotch.io/documentation/getting-started/rest/uploading-data/

Reply:
This is the error I see when I attempt to upload in `Hoppscotch` using `multipart/form-data`:
<img width="1421" alt="image" src="https://github.com/dwyl/imgup/assets/194400/93f20bd6-99e3-4701-b2ac-90e72376ba3e">

I don't _understand_ how/why you thought it was "OK" to half-ass this section of the docs. 🤷‍♂️ (😢)

Reply:
Documented in `api.md` ✅ 

<img width="1134" alt="imgup-hoppscotch-api-upload" src="https://github.com/dwyl/imgup/assets/194400/c6d529bf-effd-40b4-9a5d-e5760c199344">